docs/node-configuration.md
## Node Configuration

### Command line arguments

List of all command line arguments supported by the bundler.

```bash
Usage: cli [options]

Options:
  -V, --version                  output the version number
  --unsafe                       UNSAFE mode: Enable no storage or opcode checks during userOp simulation. SAFE mode(default).
  --network <string>             Ethereum network provider. (default: "http://localhost:8545")
  --httpApi <string>             ERC4337 rpc method namespaces to enable. (default: "web3,eth")
  --port <number>                Bundler node listening port. (default: "4337")
  --minBalance <string>          Minimum ETH balance needed for signer address. (default: "1")
  --minStake <string>            Minimum stake an entity has to have to pass the reputation system. (default: "1")
  --minUnstakeDelay <number>     Time paymaster has to wait to unlock the stake(seconds). (default: "0")
  --bundleSize <number>          Maximum number of pending mempool entities to start auto bundler. (default: "10")
  --maxBundleGas <number>        Max gas the bundler will use in transactions. (default: "5000000")
  --auto                         Automatic bundling. (default: false)
  --autoBundleInterval <number>  Auto bundler interval in (ms). (default: "12000")
  --txMode <string>              Bundler transaction mode (base, searcher).
          (base mode): Sends bundles using eth_sendRawTransaction RPC(does not protect against front running).
          (searcher mode): Sends bundles  using Flashbots Auction to protect the transaction against front running (only available on Mainnet) (default: "base")
  --metrics                      Bundler node metrics tracking enabled. (default: false)
  --influxdbUrl <string>         Url influxdb is running on (requires --metrics to be enabled). (default: "http://localhost:8086")
  --influxdbOrg <string>         Influxdb org (requires --metrics to be enabled). (default: "transeptor-labs")
  --influxdbBucket <string>      Influxdb bucket (requires --metrics to be enabled). (default: "transeptor_metrics")
  --p2p                          p2p mode enabled (default: false)
  --findPeers                    Search for peers when p2p is enabled. (default: false)
  --eip7702Support               On networks with EIP-7702 enabled, the eth_sendUserOperation method accepts an extra eip7702Auth
                                 parameter. (default: true)
  --auditTrail <boolean>         Enable audit trail. (default: false)
  -h, --help                     display help for command
```

### Environment variables

List of all environment variables supported by the node.

```bash
# Required for production
TRANSEPTOR_ENTRYPOINT_ADDRESS=0x
TRANSEPTOR_BENEFICIARY=<address_to_receive_funds>
TRANSEPTOR_MNEMONIC=<your-mnemonic>

# Optional
TRANSEPTOR_INFLUX_TOKEN=DEV_TOKEN
TRANSEPTOR_WHITE_LIST=<address_to_whitelist_SEPARATED_BY_COMMA>
TRANSEPTOR_BLACK_LIST=<address_to_blacklist_SEPARATED_BY_COMMA>
```


CHANGELOG.md
# Changelog

All notable changes to this project will be manually documented in this file by the project maintainers.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [v0.13.0-alpha.0] - 2025-07-04

### Added

- Add check prerequisites script to improve developer experience ([#146](https://github.com/transeptorlabs/transeptor-bundler/pull/146))
- Introduce the Safe Module Pattern ([#147](https://github.com/transeptorlabs/transeptor-bundler/pull/147))
- Add userOp lifecycle audit trail logs ([#149](https://github.com/transeptorlabs/transeptor-bundler/pull/149))
  - Only runs audit trail in production
  - Redacts the userOp `signature` , `callData` , `factoryData` and `eip7702Aut` from logs to avoid logging sensitive data about the userOp intent.
- Added `AuditLogQueue` module - Bounded Buffer with Strict Backpressure ([#149](https://github.com/transeptorlabs/transeptor-bundler/pull/149))
- Added `AuditLogWriter` factory to allow for different logger writers. Default to `pino` write to a .log file but can be extended in the future to allow other writers types(e.g Postgress, AWS S3, etc) ([#149](https://github.com/transeptorlabs/transeptor-bundler/pull/149))
- Introduce OCAPS module to: Enforce Capability-Based Access Control on StateService calls ([#150](https://github.com/transeptorlabs/transeptor-bundler/pull/150))

### Changes

- Move to a dependency-injected logger instead of a global logger model ([#149](https://github.com/transeptorlabs/transeptor-bundler/pull/149))

## [v0.12.0-alpha.0] - 2025-06-01

### Added

- Support for EntryPoint release`v0.8` ([#136](https://github.com/transeptorlabs/transeptor-bundler/pull/136))
- Support for [EIP-7702](https://eips.ethereum.org/EIPS/eip-7702) ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- Add `Erc7562Parser` to enforce [ERC-7562](https://eips.ethereum.org/EIPS/eip-7562) rules ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- CLI flag `--eip7702Support`: enables eip7702Auth parameter for eth_sendUserOperation on networks with EIP-7702 enabled ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- Support for sending EIP-7702 Auth Type 4 transactions ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- `Docs contribution` section added to project documentation ([#144](https://github.com/transeptorlabs/transeptor-bundler/pull/144))

### Changes

- Updated `eth-node` script to use the[accountabstraction/geth-with-erc7562-tracer](https://hub.docker.com/r/accountabstraction/geth-with-erc7562-tracer) Docker image ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- Added [@ethereumjs](https://github.com/ethereumjs/ethereumjs-monorepo/tree/master/packages/tx#eoa-code-transaction-eip-7702) dependency to prepare EIP-7702 transactions ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- Update PreVerificationGasCalculator calculations ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))

### Removed

- Drops CLI flags `--tracerRpcUrl` ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))
- Drops support for legacy JS tracer/parser ([#140](https://github.com/transeptorlabs/transeptor-bundler/pull/140))

### Fixes

- Fixed multiple failing EntryPoint v0.8 spec tests ([#137](https://github.com/transeptorlabs/transeptor-bundler/pull/137))

## [v0.11.0-alpha.0] - 2025-05-02

### Changed

- Update to use node version to 22.14.0 LTS, drop [esbuild](https://esbuild.github.io/) dependency and [pkgroll](https://github.com/privatenumber/pkgroll) as the the recommended bundler for projects using tsx.([#126](https://github.com/transeptorlabs/transeptor-bundler/pull/126))
- Move cli binary and gracefully shutting down rpc server.([#127](https://github.com/transeptorlabs/transeptor-bundler/pull/127))
- Improve test coverage in rpc module([#130](https://github.com/transeptorlabs/transeptor-bundler/pull/130))
- Refactor bundle module to reduce complexity([#132](https://github.com/transeptorlabs/transeptor-bundler/pull/130))

## [v0.10.0-alpha.0] - 2025-03-29

### Added

- A docs for Release process and Vision([#109](https://github.com/transeptorlabs/transeptor-bundler/pull/104))
- Use [esbuild](https://esbuild.github.io/) to create build for transeptor([#117](https://github.com/transeptorlabs/transeptor-bundler/pull/110))
- Add typedocs([#118](https://github.com/transeptorlabs/transeptor-bundler/pull/104))

### Changed

- Refactor modules to improve error handling with Either monad([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115)):
  - sim
  - validation
- Move all types under `../src/types`([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115))
- Configure test coverage report in `vitest.config.ts`([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115))
- Move all test to `./test` to separate test files from implementation. Moving forward, all test cases will be defined here.([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115))
- A `@vitest/coverage-istanbul:v3.0.5` to `devDependencies` to produce a test coverage report.([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115))
- Bump `devDependencies`([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115)):
  - `vite` from 5.3.1 to 5.4.12
  - `vitest` from 1.6.0 to 3.0.5
- Fix a few circular dependencies([#115](https://github.com/transeptorlabs/transeptor-bundler/pull/115))
- Drop `ts-node` and `nodemon` to use [tsx](https://tsx.is/) for local ts dev scripts([#118](https://github.com/transeptorlabs/transeptor-bundler/pull/118))
- Refactor `rpc` module to improve modularity([#119](https://github.com/transeptorlabs/transeptor-bundler/pull/119))

## [v0.9.0-alpha.0] - 2025-01-02

### Added


src/config/Config.ts
import { Command, OptionValues } from 'commander'
import dotenv from 'dotenv'
import { BigNumber, Wallet, ethers, providers } from 'ethers'
import { parseEther } from 'ethers/lib/utils.js'

import packageJson from '../../package.json' assert { type: 'json' }
import { IENTRY_POINT_ABI } from '../abis/index.js'
import { Logger } from '../logger/index.js'
import { InfluxdbConnection } from '../types/index.js'
import { isValidAddress } from '../utils/index.js'
dotenv.config()

export class Config {
  private DEFAULT_NETWORK = 'http://localhost:8545' 
  private DEFAULT_ENTRY_POINT = '0x0000000071727De22E5E9d8BAf0edAc6f37da032'
  private SUPPORTED_MODES = ['base', 'conditional', 'searcher']
  private SUPPORTED_NAMESPACES = ['web3', 'eth', 'debug']

  public readonly provider: providers.JsonRpcProvider
  public readonly connectedWallet: Wallet
  public readonly beneficiaryAddr: string
  public readonly entryPointContract: ethers.Contract

  public readonly autoBundleInterval: number
  public readonly bundleSize: number
  public readonly isAutoBundle: boolean
  public readonly maxMempoolSize: number = 100

  public readonly minStake: BigNumber
  public readonly minUnstakeDelay: number

  public readonly minSignerBalance: BigNumber
  public readonly maxBundleGas: number

  public readonly port: number
  public readonly txMode: string
  public readonly clientVersion: string
  public readonly isUnsafeMode: boolean
  public readonly isP2PMode: boolean
  public readonly findPeers: boolean

  public readonly whitelist: string[]
  public readonly blacklist: string[]
  public readonly peerMultiaddrs: string[]

  public readonly httpApi: string[]

  public readonly isMetricsEnabled: boolean
  readonly metricsPort: number
  public readonly influxdbConnection: InfluxdbConnection  = {
    url: '',
    org: '',
    bucket: '',
    token: ''
  }

  constructor(args: readonly string[]) {
    const program = new Command()
    program
    .version(`${packageJson.version}`)
    .option('--httpApi <string>', 'ERC4337 rpc method name spaces to enable.', 'web3,eth')
    .option('--network <string>', 'ETH execution client url.', `${this.DEFAULT_NETWORK}`)
    .option('--minBalance <string>', 'Maximum ETH balance need for signer address.', '1')
    .option('--maxBundleGas <number>', 'Max gas the bundler will use in transactions.', '5000000')
    .option('--auto', 'Automatic bundling.', false)
    .option('--autoBundleInterval <number>', 'Auto bundler interval in (ms).', '12000')
    .option('--bundleSize <number>', 'Maximum number of pending mempool entities to start auto bundler.', '10')
    .option('--port <number>', 'Bundler node listening port.', '4337')
    .option('--minStake <string>', 'Minimum stake a entity has to have to pass reputation system.', '1') // The stake value is not enforced on-chain, but specifically by each node while simulating a transaction
    .option('--minUnstakeDelay <number>', 'Time paymaster has to wait to unlock the stake(seconds).', '0') // One day - 84600
    .option('--txMode <string>', 'Bundler transaction mode (base, conditional, searcher).', 'base')
    .option('--unsafe', 'Enable no storage or opcode checks.')
    .option('--p2p', 'p2p mode enabled', false)
    .option('--findPeers', 'Search for peers when p2p enabled.', false)
    .option('--metrics', 'Bundler node metrics tracking enabled.', false)
    .option('--metricsPort <number>', 'Metrics server listening port.', '4001')
    .option('--influxdbUrl <string>', 'Url influxdb is running on (requires --metrics to be enabled).', 'http://localhost:8086')
    .option('--influxdbOrg <string>', 'Influxdb org (requires --metrics to be enabled).', 'transeptor-labs')
    .option('--influxdbBucket <string>', 'Influxdb bucket (requires --metrics to be enabled).', 'transeptor_metrics')

    const programOpts: OptionValues = program.parse(args).opts()
        
    if (this.SUPPORTED_MODES.indexOf(programOpts.txMode as string) === -1) {      
      throw new Error('Invalid bundler mode')
    }

    if (programOpts.txMode as string === 'searcher') {
      if (!process.env.TRANSEPTOR_ALCHEMY_API_KEY) {
        throw new Error('TRANSEPTOR_ALCHEMY_API_KEY env var not set')
      }
      
      this.provider = this.getNetworkProvider(programOpts.network as string, process.env.TRANSEPTOR_ALCHEMY_API_KEY as string)
    } else {
      this.provider = this.getNetworkProvider(programOpts.network as string)
    } 

    let supportedEntryPointAddress: string
    if (!process.env.TRANSEPTOR_ENTRYPOINT_ADDRESS) {
      supportedEntryPointAddress = this.DEFAULT_ENTRY_POINT
    } else {
      supportedEntryPointAddress = process.env.TRANSEPTOR_ENTRYPOINT_ADDRESS as string
    }
    if (!isValidAddress(supportedEntryPointAddress)) {
      throw new Error('Entry point not a valid address')
    }

    if (!process.env.TRANSEPTOR_MNEMONIC) {
      throw new Error('TRANSEPTOR_MNEMONIC env var not set')
    }

    if (!process.env.TRANSEPTOR_BENEFICIARY) {
      throw new Error('TRANSEPTOR_BENEFICIARY env var not set')
    }
    if (!isValidAddress(process.env.TRANSEPTOR_BENEFICIARY as string)) {
      throw new Error('Beneficiary not a valid address')
    }

    if(!process.env.WHITELIST) {
      this.whitelist = []
    } else {
      this.whitelist = process.env.WHITELIST.split(',')
    }

    if(!process.env.BLACKLIST) {
      this.blacklist = []
    } else {
      this.blacklist = process.env.BLACKLIST.split(',')
    }

    this.isP2PMode = programOpts.p2p as boolean
    if (this.isP2PMode) {
      if(!process.env.PEER_MULTIADDRS) {
        this.peerMultiaddrs = []
      } else {
        this.peerMultiaddrs = process.env.PEER_MULTIADDRS.split(',')
      }
    } else {
      this.peerMultiaddrs = []
    }

    this.isMetricsEnabled = programOpts.metrics as boolean
    if (this.isMetricsEnabled) {
      if (!process.env.TRANSEPTOR_INFLUX_TOKEN) {
        throw new Error('TRANSEPTOR_INFLUX_TOKEN env var not set')
      }
      this.influxdbConnection = {
        url: programOpts.influxdbUrl as string,
        token: process.env.TRANSEPTOR_INFLUX_TOKEN as string,
        org: programOpts.influxdbOrg as string,
        bucket: programOpts.influxdbBucket as string
      }
      this.metricsPort = parseInt(programOpts.metricsPort as string)
      Logger.info(`Metrics enabled, connecting to influxdb at ${this.influxdbConnection.url} with org ${this.influxdbConnection.org} and bucket ${this.influxdbConnection.bucket}`)
    }
  
    this.connectedWallet = Wallet.fromMnemonic(process.env.TRANSEPTOR_MNEMONIC as string).connect(this.provider)
    this.beneficiaryAddr = process.env.TRANSEPTOR_BENEFICIARY as string
    this.entryPointContract = new ethers.Contract(supportedEntryPointAddress, IENTRY_POINT_ABI, this.connectedWallet)

    this.autoBundleInterval = parseInt(programOpts.autoBundleInterval as string)
    this.bundleSize = parseInt(programOpts.bundleSize as string)
    this.isAutoBundle = programOpts.auto as boolean

    this.minStake = parseEther(programOpts.minStake as string)
    this.minUnstakeDelay = parseInt(programOpts.minUnstakeDelay as string)

    this.minSignerBalance = parseEther(programOpts.minBalance as string)
    this.maxBundleGas = parseInt(programOpts.maxBundleGas as string)

    this.port = parseInt(programOpts.port as string)
    this.txMode = programOpts.txMode as string
    this.clientVersion = packageJson.version as string
    this.isUnsafeMode = programOpts.unsafe as boolean ? true : false
    this.findPeers = programOpts.findPeers as boolean ? true : false

    this.httpApi = (programOpts.httpApi as string).split(',')
    for (let i = 0; i < this.httpApi.length; i++) {
      if (this.SUPPORTED_NAMESPACES.indexOf(this.httpApi[i]) === -1) {
        throw new Error('Invalid http api')
      }
    }

    this.dump()
  }

  private getNetworkProvider(url: string, apiKey?: string): providers.JsonRpcProvider {
    const isValid = this.isValidUrl(url)
    if (!isValid) {
      throw new Error('Invalid network url')
    }
    return apiKey ? new providers.JsonRpcProvider(`${url.replace(/\/+$/, '')}/${apiKey}`) : new providers.JsonRpcProvider(url)
  }

  private isValidUrl(url: string): boolean {
    const pattern = /^(https?|ftp):\/\/[^\s/$.?#].[^\s]*$/
    return pattern.test(url)
  }

  private dump(): void {
    Logger.debug(
      {
      clientVersion: this.clientVersion,
      txMode: this.txMode,
      isUnsafeMode: this.isUnsafeMode,
      isP2PMode: this.isP2PMode,
      peerMultiaddrs: this.peerMultiaddrs,
      httpApi: this.httpApi,
      beneficiaryAddr: this.beneficiaryAddr,
      entryPointContractAddress: this.entryPointContract.address,
      autoBundleInterval: this.autoBundleInterval,
      bundleSize: this.bundleSize,
      isAutoBundle: this.isAutoBundle,
      minStake: `${this.minStake.toString()} wei`,
      minUnstakeDelay: this.minUnstakeDelay.toString(),
      minSignerBalance: `${this.minSignerBalance.toString()} wei`,
      maxBundleGas: this.maxBundleGas,
      port: this.port,
      whitelist: this.whitelist,
      blacklist: this.blacklist,
      metrics: this.isMetricsEnabled,
      metricsPort: this.metricsPort,
    },
    'Bundler config setup')
  }
}
